In [1]:
import numpy as np
import pandas as pd
from itertools import product
from sklearn.preprocessing import MinMaxScaler
import tensorflow as tf
from tensorflow import keras
import random
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from sklearn.metrics import mean_absolute_percentage_error, mean_absolute_error, mean_squared_error

2025-01-09 04:16:34.181387: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-01-09 04:16:34.195672: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2025-01-09 04:16:34.212366: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2025-01-09 04:16:34.217252: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-01-09 04:16:34.229907: I tensorflow/core/platform/cpu_feature_guar

In [2]:
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))
tf.config.list_physical_devices()

Num GPUs Available:  1


[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [3]:
# List all physical devices
gpus = tf.config.list_physical_devices('GPU')
print("Available GPUs:")
for gpu in gpus:
    print(f" - {gpu.name}")

Available GPUs:
 - /physical_device:GPU:0


In [4]:
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        print("Memory growth enabled for GPUs.")
    except RuntimeError as e:
        print(e)

Memory growth enabled for GPUs.


In [5]:
traffic_data = pd.read_csv('GD030A_S.csv')

## 1. Recover timestamp

In [6]:
# Define the recover_timestamp function
def recover_timestamp(data):
    # Combine 'date' and 'time' to form a datetime column
    data['datetime'] = pd.to_datetime(data['date'] + ' ' + data['time'].astype(str) + ':00', format='%Y-%m-%d %H:%M')

    # Set 'datetime' as index
    data = data.set_index('datetime')

    # Create a complete range of timestamps with hourly frequency
    full_time_range = pd.date_range(start=data.index.min(), end=data.index.max(), freq='H')

    # Reindex the data to include all timestamps, filling missing rows with NaN
    data_full = data.reindex(full_time_range)

    return data_full

In [7]:
# Apply the recover_timestamp function to recover the full time series
traffic_full = recover_timestamp(traffic_data)
traffic_full

date  time   flow
2019-10-01 00:00:00  2019-10-01   0.0   15.0
2019-10-01 01:00:00  2019-10-01   1.0    9.0
2019-10-01 02:00:00  2019-10-01   2.0    9.0
2019-10-01 03:00:00  2019-10-01   3.0    7.0
2019-10-01 04:00:00  2019-10-01   4.0    9.0
...                         ...   ...    ...
2023-09-30 19:00:00  2023-09-30  19.0  129.0
2023-09-30 20:00:00  2023-09-30  20.0  119.0
2023-09-30 21:00:00  2023-09-30  21.0  106.0
2023-09-30 22:00:00  2023-09-30  22.0   88.0
2023-09-30 23:00:00  2023-09-30  23.0   88.0

[35064 rows x 3 columns]

## 2. Train, validate, test data split

In [8]:
train_set = traffic_full[:'2022-02-28 23:00:00']
valid_set = traffic_full['2022-03-01 00:00:00':'2022-12-31 23:00:00']
test_set = traffic_full['2023-01-01 00:00:00':]
print('Proportion of train_set : {:.4f}'.format(len(train_set)/len(traffic_full)))
print('Proportion of valid_set : {:.4f}'.format(len(valid_set)/len(traffic_full)))
print('Proportion of test_set : {:.4f}'.format(len(test_set)/len(traffic_full)))

Proportion of train_set : 0.6037
Proportion of valid_set : 0.2094
Proportion of test_set : 0.1869


In [9]:
print(train_set.isnull().sum(), len(train_set))
print(valid_set.isnull().sum(),len(valid_set))
print(test_set.isnull().sum(),len(test_set))

date    927
time    927
flow    927
dtype: int64 21168
date    91
time    91
flow    91
dtype: int64 7344
date    403
time    403
flow    403
dtype: int64 6552


## 3. Normalise the data 

In [10]:
# Initialize the scaler
scaler = MinMaxScaler()

# Fit the scaler on the training data's 'flow' feature
scaler.fit(train_set[['flow']])

# Transform the 'flow' feature in all datasets
train_set.loc[:, 'flow_scaled'] = scaler.transform(train_set[['flow']])
valid_set.loc[:, 'flow_scaled'] = scaler.transform(valid_set[['flow']])
test_set.loc[:, 'flow_scaled'] = scaler.transform(test_set[['flow']])

/tmp/ipykernel_4038141/1100395100.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_set.loc[:, 'flow_scaled'] = scaler.transform(train_set[['flow']])
/tmp/ipykernel_4038141/1100395100.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  valid_set.loc[:, 'flow_scaled'] = scaler.transform(valid_set[['flow']])
/tmp/ipykernel_4038141/1100395100.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See th

## 4. Split the data into X and y

In [11]:
def create_sequences(data, input_length, forecast_horizon):
    """
    Creates input-output sequences for time series data, excluding any sequences containing NaN values.
    
    Parameters:
    - data: pandas DataFrame containing the data. Must include the 'flow_scaled' column.
    - input_length: int, number of past time steps to include in each input sequence.
    - forecast_horizon: int, number of future steps to predict.
    
    Returns:
    - X: numpy array of shape (num_valid_samples, input_length, num_features)
    - y: numpy array of shape (num_valid_samples, forecast_horizon)
    """
    X, y = [], []
    num_features = data.shape[1]
    total_length = input_length + forecast_horizon
    
    for i in range(input_length, len(data) - forecast_horizon + 1):
        # Extract the input sequence
        X_seq = data.iloc[i - input_length:i]['flow_scaled'].values
        # Extract the target sequence
        y_seq = data.iloc[i:i + forecast_horizon]['flow_scaled'].values
        
        # Check for NaN values in the input sequence and target sequence
        if not np.isnan(X_seq).any() and not np.isnan(y_seq).any():
            X.append(X_seq)
            y.append(y_seq)
        else:
            # Optionally, log or count the skipped sequences
            pass  # Simply skip sequences with NaNs
        
    # Convert to numpy arrays and reshape X to match LSTM expected input (samples, timesteps, features)
    X = np.array(X).reshape(-1, input_length, 1)
    y = np.array(y).reshape(-1, forecast_horizon)
    
    return X, y

## 5. Create X and y

In [12]:
# Define Input Sequence Lengths
input_lengths = [24 * i for i in range(1, 22)]  # [24, 48, ..., 168]

In [13]:
from collections import defaultdict
data_dict = defaultdict(dict)

for length in input_lengths:
    print(f"Processing input length: {length}")
    
    # Create sequences with forecast_horizon=6
    X_train, y_train = create_sequences(train_set, length, forecast_horizon=6)
    X_val, y_val = create_sequences(valid_set, length, forecast_horizon=6)
    X_test, y_test = create_sequences(test_set, length, forecast_horizon=6)
    
    # Store in the dictionary
    data_dict[length]['X_train'] = X_train
    data_dict[length]['y_train'] = y_train
    data_dict[length]['X_val'] = X_val
    data_dict[length]['y_val'] = y_val
    data_dict[length]['X_test'] = X_test
    data_dict[length]['y_test'] = y_test
    
    # Print shapes and ensure no NaNs
    print(f"  X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
    print(f"  X_val shape: {X_val.shape}, y_val shape: {y_val.shape}")
    print(f"  X_test shape: {X_test.shape}, y_test shape: {y_test.shape}\n")

Processing input length: 24
  X_train shape: (19806, 24, 1), y_train shape: (19806, 6)
  X_val shape: (7108, 24, 1), y_val shape: (7108, 6)
  X_test shape: (5915, 24, 1), y_test shape: (5915, 6)

Processing input length: 48
  X_train shape: (19446, 48, 1), y_train shape: (19446, 6)
  X_val shape: (6988, 48, 1), y_val shape: (6988, 6)
  X_test shape: (5728, 48, 1), y_test shape: (5728, 6)

Processing input length: 72
  X_train shape: (19086, 72, 1), y_train shape: (19086, 6)
  X_val shape: (6868, 72, 1), y_val shape: (6868, 6)
  X_test shape: (5560, 72, 1), y_test shape: (5560, 6)

Processing input length: 96
  X_train shape: (18726, 96, 1), y_train shape: (18726, 6)
  X_val shape: (6748, 96, 1), y_val shape: (6748, 6)
  X_test shape: (5392, 96, 1), y_test shape: (5392, 6)

Processing input length: 120
  X_train shape: (18366, 120, 1), y_train shape: (18366, 6)
  X_val shape: (6628, 120, 1), y_val shape: (6628, 6)
  X_test shape: (5224, 120, 1), y_test shape: (5224, 6)

Processing input

## 6. Build LSTM model

In [14]:
def build_improved_lstm_model(hyperparams, input_length):
    """
    Builds an improved LSTM model based on provided hyperparameters and input length.

    Parameters:
    - hyperparams: dict containing 'units', 'dropout', 'learning_rate', and optionally 'recurrent_dropout'.
    - input_length: int, length of the input sequences.

    Returns:
    - model: compiled Keras model ready for training.
    """
    model = Sequential()
    
    # First LSTM layer with return_sequences=True to stack another LSTM layer
    model.add(LSTM(
        units=hyperparams['units'], 
        activation='tanh', 
        input_shape=(input_length, 1), 
        return_sequences=True, 
        #recurrent_dropout=hyperparams.get('recurrent_dropout', 0.0)
    ))
    model.add(Dropout(rate=hyperparams['dropout']))
    
    # Second LSTM layer
    model.add(LSTM(
        units=hyperparams['units'], 
        activation='tanh',
        #recurrent_dropout=hyperparams.get('recurrent_dropout', 0.0)
    ))
    model.add(Dropout(rate=hyperparams['dropout']))
    
    # Output layer for multi-step forecasting
    model.add(Dense(6))
    
    # Compile the model with MSE as the loss function
    optimizer = keras.optimizers.Adam(learning_rate=hyperparams['learning_rate'])
    model.compile(optimizer=optimizer, loss='mse', metrics=['mse'])
    
    return model


## 7. Retrain the model after getting the best hyperparameters of each input length

In [15]:
best_hyperparameters_improved_model_lstm = {
    24: {
        'units': 64,
        'dropout': 0.2,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    48: {
        'units': 256,
        'dropout': 0.2,
        'learning_rate': 0.01,
        'batch_size': 32
    },
    72: {
        'units': 256,
        'dropout': 0.1,
        'learning_rate': 0.001,
        'batch_size': 128
    },
    96: {
        'units': 256,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    120: {
        'units': 128,
        'dropout': 0.1,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    144: {
        'units': 256,
        'dropout': 0.2,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    168: {
        'units': 256,
        'dropout': 0.3,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    192: {
        'units': 256,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    216: {
        'units': 256,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    240: {
        'units': 64,
        'dropout': 0.4,
        'learning_rate': 0.005,
        'batch_size': 64
    },
    264: {
        'units': 256,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    288: {
        'units': 64,
        'dropout': 0.3,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    312: {
        'units': 32,
        'dropout': 0.2,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    336: {
        'units': 256,
        'dropout': 0.1,
        'learning_rate': 0.01,
        'batch_size': 64
    },
    360: {
        'units': 32,
        'dropout': 0.1,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    384: {
        'units': 256,
        'dropout': 0.3,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    408: {
        'units': 256,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    432: {
        'units': 128,
        'dropout': 0.4,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    456: {
        'units': 128,
        'dropout': 0.0,
        'learning_rate': 0.001,
        'batch_size': 32
    },
    480: {
        'units': 256,
        'dropout': 0.1,
        'learning_rate': 0.001,
        'batch_size': 64
    },
    504: {
        'units': 128,
        'dropout': 0.0,
        'learning_rate': 0.001,
        'batch_size': 32
    }
}


In [16]:
import logging

logging.basicConfig(
    level=logging.INFO,  # Set the logging level to INFO
    format='%(asctime)s - %(message)s',  # Customize the log message format
    handlers=[
        logging.FileHandler('lstm2.log'),  # Log messages to 'output.log'
        logging.StreamHandler()             # Also output to console/notebook
    ]
)

In [17]:
def set_seed(seed):
    np.random.seed(seed)
    random.seed(seed)
    tf.random.set_seed(seed)

In [18]:
def train_model(hyperparams,data_dict,length, seed=None):  # add seed
    if seed is not None:
        set_seed(seed)
    
    #get the data of each length
    X_train = data_dict[length]['X_train']
    y_train = data_dict[length]['y_train']
    X_val = data_dict[length]['X_val']
    y_val = data_dict[length]['y_val']
    
    # Train the model
    #model = build_lstm_model(hyperparams, length)    
    model = build_improved_lstm_model(hyperparams, length)
    # Early Stopping Callback
    early_stop = EarlyStopping(monitor='val_loss', patience=20, restore_best_weights=True,verbose=1)    
    
    history = model.fit(
        X_train, y_train,
        epochs=100,
        batch_size=hyperparams['batch_size'],
        validation_data=(X_val, y_val),
        callbacks=[early_stop],
        verbose=0  # Set to 1 to see training progress
    )
    
    # Retrieve the best validation MSE from the history
    best_mse = min(history.history['val_loss'])
    logging.info(f"Validation loss: {best_mse:.5f}")
    
    return model, best_mse

In [19]:
# Make predictions
def make_prediction(model, X_obs, y_obs):
    y_pred = model.predict(X_obs,verbose=0)
    n_samples = X_obs.shape[0]
    output_len = y_obs.shape[1]

    # Reshape for inverse scaling
    y_pred_reshaped = y_pred.reshape(-1, 1)
    y_obs_reshaped = y_obs.reshape(-1, 1)

    # Inverse transform
    y_pred_inverse = scaler.inverse_transform(y_pred_reshaped).reshape(n_samples, output_len)
    y_obs_inverse = scaler.inverse_transform(y_obs_reshaped).reshape(n_samples, output_len)

    return y_pred_inverse, y_obs_inverse

In [20]:
# Compute Metrics for Each Time Step
def evaluation(y_pred_inverse, y_obs_inverse):
    
    output_len = y_pred_inverse.shape[1]
    metrics_list = []  # To store metrics for each time step
    
    for i in range(output_len):
        y_true = y_obs_inverse[:, i]
        y_pred = y_pred_inverse[:, i]

        # Mean Absolute Error (MAE)
        mae = mean_absolute_error(y_true, y_pred)

        # Mean Squared Error (MSE)
        mse = mean_squared_error(y_true, y_pred)

        # Root Mean Squared Error (RMSE)
        rmse = np.sqrt(mse)

        # Mean Absolute Percentage Error (MAPE)
        # Avoid division by zero by adding a small epsilon to y_test_flat if necessary
        epsilon = 1e-10
        y_true_safe = np.where(y_true == 0, epsilon, y_true)
        mape = np.mean(np.abs((y_true - y_pred) / y_true_safe)) * 100

                # Append the metrics for the current time step to the list
        metrics_list.append({
            'Time Step': i + 1,
            'MAE': mae,
            'RMSE': rmse,
            'MAPE (%)': mape
        })

    # Create a DataFrame from the list of metrics
    metrics_df = pd.DataFrame(metrics_list)
    metrics_df.set_index('Time Step', inplace=True)

    return metrics_df

In [ ]:
mean_metrics_list=[]
#for length in best_hyperparameters_improved_model.keys():
for length in best_hyperparameters_improved_model_lstm.keys():
    logging.info(length)
    
    # Number of runs
    n_runs = 10
    
    # get the best hyperparameter of each length
    #hyperparams = best_hyperparameters_improved_model[length]
    hyperparams = best_hyperparameters_improved_model_lstm[length]
    
    # Initialize lists to store mean metrics and all metrics from each run
    mean_metrics_rows = []
    df_all_metrics_list = []
    successful_runs = []  # To track runs with valid mse

    for run in range(1, n_runs + 1):
        logging.info(f"\n--- Run {run} ---")

        # Optionally set a unique seed for each run to ensure variability
        seed = run
        # Train the model
        model, mse = train_model(hyperparams, data_dict, length, seed=seed)
        
        if not np.isnan(mse):  # Only proceed if mse is valid
            successful_runs.append(run)
        
            X_train = data_dict[length]['X_train']
            y_train = data_dict[length]['y_train']
            X_val = data_dict[length]['X_val']
            y_val = data_dict[length]['y_val']
            X_test = data_dict[length]['X_test']
            y_test = data_dict[length]['y_test']

            #get the true flow and predicted flow
            y_pred_train, y_obs_train = make_prediction(model, X_train, y_train)
            y_pred_val, y_obs_val = make_prediction(model, X_val, y_val)
            y_pred_test, y_obs_test = make_prediction(model, X_test, y_test)

            #calculate the evaluation metrics of each output step
            df_train = evaluation(y_pred_train, y_obs_train).add_suffix('_train')
            df_val = evaluation(y_pred_val, y_obs_val).add_suffix('_val')
            df_test = evaluation(y_pred_test, y_obs_test).add_suffix('_test')

            df_all_metrics = pd.concat([df_train, df_val, df_test], axis=1)
            df_all_metrics.index.name = length

            # Append df_all_metrics to the list
            df_all_metrics_list.append(df_all_metrics)

            # Calculate mean for all output step
            #mean_metrics = df_val.mean()
            mean_metrics = pd.concat([df_train.mean(), df_val.mean(), df_test.mean()])
            mean_metrics_row = pd.DataFrame(mean_metrics).T
            mean_metrics_row['MSE_val(loss)'] = mse
            mean_metrics_row['input_len'] = length
            #mean_metrics_row = mean_metrics_row[['input_len','MSE_val(loss)', 'MAE_val', 'RMSE_val', 'MAPE (%)_val']]
            mean_metrics_row = mean_metrics_row[['input_len', 'MSE_val(loss)', 'MAE_train', 'RMSE_train', 'MAPE (%)_train',
                                             'MAE_val', 'RMSE_val', 'MAPE (%)_val',
                                             'MAE_test', 'RMSE_test', 'MAPE (%)_test']]

            # Append to the list
            mean_metrics_rows.append(mean_metrics_row)
        else:
            logging.info(f"Run {run} has mse=NaN, skipping.")
            
    # Check if there are successful runs before proceeding
    if successful_runs:       
        # Concatenate all df_all_metrics into a single DataFrame with a new level for runs
        concatenated_all_metrics = pd.concat(df_all_metrics_list, keys=successful_runs, names=['Run', 'Time Step'])

        # Calculate the mean across runs for each metric and time step
        # This will group by 'Time Step' and calculate the mean of each metric across all runs
        aggregated_all_metrics_mean = concatenated_all_metrics.groupby('Time Step').mean()
        aggregated_all_metrics_mean.index.name = length

        logging.info("\n--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---")
        # Convert DataFrame to string
        aggregated_all_metrics_mean_str = aggregated_all_metrics_mean.to_string(index=False)    
        # Log the DataFrame
        logging.info("\n" + aggregated_all_metrics_mean_str)
        display(aggregated_all_metrics_mean)
    else:
        logging.info(f"No successful runs for input length {length}.")

    # Check if mean metrics were calculated
    if mean_metrics_rows:        
        # After all runs, create a DataFrame of mean metrics
        mean_metrics_df = pd.concat(mean_metrics_rows, ignore_index=True)
        # Calculate the mean of each metric across the 10 runs
        final_mean_metrics = mean_metrics_df.mean()

        # Create a DataFrame for the final mean metrics
        final_mean_metrics_df = pd.DataFrame(final_mean_metrics).T

        mean_metrics_list.append(final_mean_metrics_df)
    else:
        logging.info(f"No mean metrics calculated for input length {length}.")

mean_metrics_df = pd.concat(mean_metrics_list).reset_index(drop=True)
logging.info("\n--- Final Mean Metrics Across 10 Runs ---")
# Convert DataFrame to string
mean_metrics_df_str = mean_metrics_df.to_string(index=False)    
# Log the DataFrame
logging.info("\n" + mean_metrics_df_str)
mean_metrics_df

2025-01-09 04:18:04,995 - 24
2025-01-09 04:18:04,996 - 
--- Run 1 ---
2025-01-09 04:18:05.202507: I tensorflow/core/common_runtime/gpu/gpu_device.cc:2021] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 18612 MB memory:  -> device: 0, name: NVIDIA A30, pci bus id: 276f:00:00.0, compute capability: 8.0
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)
2025-01-09 04:18:08.277144: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:531] Loaded cuDNN version 8907


Epoch 72: early stopping
Restoring model weights from the end of the best epoch: 52.


2025-01-09 04:26:46,176 - Validation loss: 0.00593
2025-01-09 04:26:49,989 - 
--- Run 2 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 86: early stopping
Restoring model weights from the end of the best epoch: 66.


2025-01-09 04:36:32,982 - Validation loss: 0.00595
2025-01-09 04:36:37,037 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 79: early stopping
Restoring model weights from the end of the best epoch: 59.


2025-01-09 04:46:48,139 - Validation loss: 0.00580
2025-01-09 04:46:52,206 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 66: early stopping
Restoring model weights from the end of the best epoch: 46.


2025-01-09 04:55:52,043 - Validation loss: 0.00574
2025-01-09 04:55:56,115 - 
--- Run 5 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 76: early stopping
Restoring model weights from the end of the best epoch: 56.


2025-01-09 05:06:25,406 - Validation loss: 0.00580
2025-01-09 05:06:29,422 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 63: early stopping
Restoring model weights from the end of the best epoch: 43.


2025-01-09 05:15:29,823 - Validation loss: 0.00578
2025-01-09 05:15:34,051 - 
--- Run 7 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 68: early stopping
Restoring model weights from the end of the best epoch: 48.


2025-01-09 05:26:08,138 - Validation loss: 0.00591
2025-01-09 05:26:12,002 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 65: early stopping
Restoring model weights from the end of the best epoch: 45.


2025-01-09 05:35:57,998 - Validation loss: 0.00581
2025-01-09 05:36:02,230 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 81: early stopping
Restoring model weights from the end of the best epoch: 61.


2025-01-09 05:48:03,354 - Validation loss: 0.00572
2025-01-09 05:48:07,641 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 66: early stopping
Restoring model weights from the end of the best epoch: 46.


2025-01-09 05:57:47,647 - Validation loss: 0.00594
2025-01-09 05:57:52,001 - 
--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---
2025-01-09 05:57:52,004 - 
 MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
 22.907106   32.069606    7.455691e+10 20.861006 32.247095     38.358877 22.904381  33.237082      26.900769
 25.212608   35.473405    1.206191e+11 23.232720 36.871149     42.251573 26.379164  38.691385      30.707837
 26.637028   37.486204    1.565179e+11 24.479816 39.286775     45.731592 28.305923  41.497392      33.981775
 27.275202   38.512002    1.757934e+11 25.045882 40.616910     47.626319 29.382252  43.273759      35.839845
 28.194124   39.816366    1.984042e+11 25.551895 41.497485     51.388960 30.282175  44.693641      38.718353
 29.125494   41.100228    2.181764e+11 26.257530 42.379002     55.193810 31.232502  46.122187      41.100327


MAE_train  RMSE_train  MAPE (%)_train    MAE_val   RMSE_val  MAPE (%)_val  \
24                                                                              
1   22.907106   32.069606    7.455691e+10  20.861006  32.247095     38.358877   
2   25.212608   35.473405    1.206191e+11  23.232720  36.871149     42.251573   
3   26.637028   37.486204    1.565179e+11  24.479816  39.286775     45.731592   
4   27.275202   38.512002    1.757934e+11  25.045882  40.616910     47.626319   
5   28.194124   39.816366    1.984042e+11  25.551895  41.497485     51.388960   
6   29.125494   41.100228    2.181764e+11  26.257530  42.379002     55.193810   

     MAE_test  RMSE_test  MAPE (%)_test  
24                                       
1   22.904381  33.237082      26.900769  
2   26.379164  38.691385      30.707837  
3   28.305923  41.497392      33.981775  
4   29.382252  43.273759      35.839845  
5   30.282175  44.693641      38.718353  
6   31.232502  46.122187      41.100327

2025-01-09 05:57:52,021 - 48
2025-01-09 05:57:52,023 - 
--- Run 1 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 37: early stopping
Restoring model weights from the end of the best epoch: 17.


2025-01-09 06:04:25,261 - Validation loss: 0.00580
2025-01-09 06:04:30,296 - 
--- Run 2 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 40: early stopping
Restoring model weights from the end of the best epoch: 20.


2025-01-09 06:11:17,690 - Validation loss: 0.00587
2025-01-09 06:11:23,076 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 34: early stopping
Restoring model weights from the end of the best epoch: 14.


2025-01-09 06:17:27,778 - Validation loss: 0.00564
2025-01-09 06:17:33,018 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 61: early stopping
Restoring model weights from the end of the best epoch: 41.


2025-01-09 06:28:25,854 - Validation loss: 0.01326
2025-01-09 06:28:31,094 - 
--- Run 5 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 48: early stopping
Restoring model weights from the end of the best epoch: 28.


2025-01-09 06:36:56,257 - Validation loss: 0.00559
2025-01-09 06:37:01,584 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 23: early stopping
Restoring model weights from the end of the best epoch: 3.


2025-01-09 06:41:00,202 - Validation loss: 0.01967
2025-01-09 06:41:05,509 - 
--- Run 7 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 53: early stopping
Restoring model weights from the end of the best epoch: 33.


2025-01-09 06:50:15,243 - Validation loss: 0.00552
2025-01-09 06:50:20,462 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 44: early stopping
Restoring model weights from the end of the best epoch: 24.


2025-01-09 06:57:59,113 - Validation loss: 0.00570
2025-01-09 06:58:04,532 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 40: early stopping
Restoring model weights from the end of the best epoch: 20.


2025-01-09 07:05:09,843 - Validation loss: 0.00567
2025-01-09 07:05:15,321 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 41: early stopping
Restoring model weights from the end of the best epoch: 21.


2025-01-09 07:12:53,073 - Validation loss: 0.00574
2025-01-09 07:12:58,647 - 
--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---
2025-01-09 07:12:58,650 - 
 MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
 29.520270   40.288884    1.094725e+11 25.828811 37.947781     60.494970 28.472036  40.253431      38.975361
 31.389985   43.060981    1.396529e+11 27.885205 41.742396     62.953143 31.541250  45.009875      42.032715
 32.694977   44.830896    1.619273e+11 28.926176 43.576025     64.275727 33.201626  47.339935      43.722349
 33.997145   45.993082    1.851534e+11 30.366076 44.874476     72.599455 34.610442  48.831394      48.552608
 35.192981   47.183121    2.055001e+11 31.635624 46.058433     80.696225 35.905973  50.284138      53.508152
 36.309691   48.234046    2.288940e+11 33.162654 47.717849     94.259818 36.835540  51.373595      60.025666


MAE_train  RMSE_train  MAPE (%)_train    MAE_val   RMSE_val  MAPE (%)_val  \
48                                                                              
1   29.520270   40.288884    1.094725e+11  25.828811  37.947781     60.494970   
2   31.389985   43.060981    1.396529e+11  27.885205  41.742396     62.953143   
3   32.694977   44.830896    1.619273e+11  28.926176  43.576025     64.275727   
4   33.997145   45.993082    1.851534e+11  30.366076  44.874476     72.599455   
5   35.192981   47.183121    2.055001e+11  31.635624  46.058433     80.696225   
6   36.309691   48.234046    2.288940e+11  33.162654  47.717849     94.259818   

     MAE_test  RMSE_test  MAPE (%)_test  
48                                       
1   28.472036  40.253431      38.975361  
2   31.541250  45.009875      42.032715  
3   33.201626  47.339935      43.722349  
4   34.610442  48.831394      48.552608  
5   35.905973  50.284138      53.508152  
6   36.835540  51.373595      60.025666

2025-01-09 07:12:58,663 - 72
2025-01-09 07:12:58,665 - 
--- Run 1 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 40: early stopping
Restoring model weights from the end of the best epoch: 20.


2025-01-09 07:15:13,772 - Validation loss: 0.00560
2025-01-09 07:15:20,179 - 
--- Run 2 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 59: early stopping
Restoring model weights from the end of the best epoch: 39.


2025-01-09 07:18:40,754 - Validation loss: 0.00556
2025-01-09 07:18:47,372 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 63: early stopping
Restoring model weights from the end of the best epoch: 43.


2025-01-09 07:22:23,048 - Validation loss: 0.00545
2025-01-09 07:22:29,482 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 69: early stopping
Restoring model weights from the end of the best epoch: 49.


2025-01-09 07:26:23,604 - Validation loss: 0.00543
2025-01-09 07:26:29,608 - 
--- Run 5 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 61: early stopping
Restoring model weights from the end of the best epoch: 41.


2025-01-09 07:29:52,714 - Validation loss: 0.00553
2025-01-09 07:29:59,162 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 54: early stopping
Restoring model weights from the end of the best epoch: 34.


2025-01-09 07:33:01,091 - Validation loss: 0.00566
2025-01-09 07:33:07,536 - 
--- Run 7 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 58: early stopping
Restoring model weights from the end of the best epoch: 38.


2025-01-09 07:36:24,595 - Validation loss: 0.00562
2025-01-09 07:36:30,805 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 52: early stopping
Restoring model weights from the end of the best epoch: 32.


2025-01-09 07:39:27,618 - Validation loss: 0.00545
2025-01-09 07:39:33,985 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 50: early stopping
Restoring model weights from the end of the best epoch: 30.


2025-01-09 07:42:21,320 - Validation loss: 0.00554
2025-01-09 07:42:27,040 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 37: early stopping
Restoring model weights from the end of the best epoch: 17.


2025-01-09 07:44:30,393 - Validation loss: 0.00572
2025-01-09 07:44:36,547 - 
--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---
2025-01-09 07:44:36,550 - 
 MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
 22.741290   32.062950    8.379221e+10 20.647859 31.950224     36.571431 22.961092  33.119989      26.276085
 25.189450   35.400165    1.335175e+11 23.202950 36.313930     43.072672 26.365270  38.486934      30.826205
 26.391765   36.994922    1.670794e+11 24.297783 38.348797     45.876725 28.260791  41.199947      33.794819
 26.882717   37.754508    1.852435e+11 24.803514 39.385878     48.100192 29.206036  42.833218      36.133349
 27.536335   38.739569    2.031431e+11 25.245049 40.248047     50.061509 30.055210  44.361006      38.259194
 28.153583   39.667801    2.189994e+11 25.763952 41.079535     52.163343 30.805425  45.693264      39.911937


MAE_train  RMSE_train  MAPE (%)_train    MAE_val   RMSE_val  MAPE (%)_val  \
72                                                                              
1   22.741290   32.062950    8.379221e+10  20.647859  31.950224     36.571431   
2   25.189450   35.400165    1.335175e+11  23.202950  36.313930     43.072672   
3   26.391765   36.994922    1.670794e+11  24.297783  38.348797     45.876725   
4   26.882717   37.754508    1.852435e+11  24.803514  39.385878     48.100192   
5   27.536335   38.739569    2.031431e+11  25.245049  40.248047     50.061509   
6   28.153583   39.667801    2.189994e+11  25.763952  41.079535     52.163343   

     MAE_test  RMSE_test  MAPE (%)_test  
72                                       
1   22.961092  33.119989      26.276085  
2   26.365270  38.486934      30.826205  
3   28.260791  41.199947      33.794819  
4   29.206036  42.833218      36.133349  
5   30.055210  44.361006      38.259194  
6   30.805425  45.693264      39.911937

2025-01-09 07:44:36,558 - 96
2025-01-09 07:44:36,559 - 
--- Run 1 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 51: early stopping
Restoring model weights from the end of the best epoch: 31.


2025-01-09 07:56:09,812 - Validation loss: 0.00549
2025-01-09 07:56:17,058 - 
--- Run 2 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 44: early stopping
Restoring model weights from the end of the best epoch: 24.


2025-01-09 08:06:14,939 - Validation loss: 0.00539
2025-01-09 08:06:22,236 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 67: early stopping
Restoring model weights from the end of the best epoch: 47.


2025-01-09 08:21:38,498 - Validation loss: 0.00548
2025-01-09 08:21:45,935 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 53: early stopping
Restoring model weights from the end of the best epoch: 33.


2025-01-09 08:33:32,318 - Validation loss: 0.00555
2025-01-09 08:33:38,758 - 
--- Run 5 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 47: early stopping
Restoring model weights from the end of the best epoch: 27.


2025-01-09 08:44:18,349 - Validation loss: 0.00557
2025-01-09 08:44:25,411 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 46: early stopping
Restoring model weights from the end of the best epoch: 26.


2025-01-09 08:54:48,591 - Validation loss: 0.00549
2025-01-09 08:54:56,085 - 
--- Run 7 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 48: early stopping
Restoring model weights from the end of the best epoch: 28.


2025-01-09 09:06:00,642 - Validation loss: 0.00570
2025-01-09 09:06:08,255 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 50: early stopping
Restoring model weights from the end of the best epoch: 30.


2025-01-09 09:18:15,729 - Validation loss: 0.00548
2025-01-09 09:18:23,157 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 47: early stopping
Restoring model weights from the end of the best epoch: 27.


2025-01-09 09:28:59,867 - Validation loss: 0.00548
2025-01-09 09:29:07,581 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Completed grid search for input length: 24 Best Validation MSE: 0.00553 Best Hyperparameters: {'units': 64, 'dropout': 0.2, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 48 Best Validation MSE: 0.00546 Best Hyperparameters: {'units': 256, 'dropout': 0.2, 'learning_rate': 0.01, 'batch_size': 32}

Completed grid search for input length: 72 Best Validation MSE: 0.00531 Best Hyperparameters: {'units': 256, 'dropout': 0.1, 'learning_rate': 0.001, 'batch_size': 128}

Completed grid search for input length: 96 Best Validation MSE: 0.00535 Best Hyperparameters: {'units': 256, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 120 Best Validation MSE: 0.00528 Best Hyperparameters: {'units': 128, 'dropout': 0.1, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 144 Best Validation MSE: 0.00528 Best Hyperparameters: {'units': 256, 'dropout': 0.2, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 168 Best Validation MSE: 0.00526 Best Hyperparameters: {'units': 256, 'dropout': 0.3, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 192 Best Validation MSE: 0.00524 Best Hyperparameters: {'units': 256, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 216 Best Validation MSE: 0.00537 Best Hyperparameters: {'units': 256, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 240 Best Validation MSE: 0.00541 Best Hyperparameters: {'units': 64, 'dropout': 0.4, 'learning_rate': 0.005, 'batch_size': 64}

Completed grid search for input length: 264 Best Validation MSE: 0.00546 Best Hyperparameters: {'units': 256, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 288 Best Validation MSE: 0.00547 Best Hyperparameters: {'units': 64, 'dropout': 0.3, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 312 Best Validation MSE: 0.00555 Best Hyperparameters: {'units': 32, 'dropout': 0.2, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 336 Best Validation MSE: 0.00556 Best Hyperparameters: {'units': 256, 'dropout': 0.1, 'learning_rate': 0.01, 'batch_size': 64}

Completed grid search for input length: 360 Best Validation MSE: 0.00570 Best Hyperparameters: {'units': 32, 'dropout': 0.1, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 384 Best Validation MSE: 0.00574 Best Hyperparameters: {'units': 256, 'dropout': 0.3, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 408 Best Validation MSE: 0.00565 Best Hyperparameters: {'units': 256, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 432 Best Validation MSE: 0.00568 Best Hyperparameters: {'units': 128, 'dropout': 0.4, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 456 Best Validation MSE: 0.00572 Best Hyperparameters: {'units': 128, 'dropout': 0.0, 'learning_rate': 0.001, 'batch_size': 32}

Completed grid search for input length: 480 Best Validation MSE: 0.00585 Best Hyperparameters: {'units': 256, 'dropout': 0.1, 'learning_rate': 0.001, 'batch_size': 64}

Completed grid search for input length: 504 Best Validation MSE: 0.00596 Best Hyperparameters: {'units': 128, 'dropout': 0.0, 'learning_rate': 0.001, 'batch_size': 32}

### Recalculate the evaluation metrics of length 336 and 48

In [22]:
mean_metrics_list=[]
#for length in best_hyperparameters_improved_model.keys():
for length in [336,48]:
    logging.info(length)
    
    # Number of runs
    n_runs = 10
    
    # get the best hyperparameter of each length
    #hyperparams = best_hyperparameters_improved_model[length]
    hyperparams = best_hyperparameters_improved_model_lstm[length]
    
    # Initialize lists to store mean metrics and all metrics from each run
    mean_metrics_rows = []
    df_all_metrics_list = []
    successful_runs = []  # To track runs with valid mse

    for run in range(1, n_runs + 1):
        logging.info(f"\n--- Run {run} ---")

        # Optionally set a unique seed for each run to ensure variability
        seed = run
        # Train the model
        model, mse = train_model(hyperparams, data_dict, length, seed=seed)
        
        if not np.isnan(mse):  # Only proceed if mse is valid

            if mse>0.007:
                logging.info(f"Run {run} has mse={mse}, skipping.")
            else:
                successful_runs.append(run)
            
                X_train = data_dict[length]['X_train']
                y_train = data_dict[length]['y_train']
                X_val = data_dict[length]['X_val']
                y_val = data_dict[length]['y_val']
                X_test = data_dict[length]['X_test']
                y_test = data_dict[length]['y_test']
    
                #get the true flow and predicted flow
                y_pred_train, y_obs_train = make_prediction(model, X_train, y_train)
                y_pred_val, y_obs_val = make_prediction(model, X_val, y_val)
                y_pred_test, y_obs_test = make_prediction(model, X_test, y_test)
    
                #calculate the evaluation metrics of each output step
                df_train = evaluation(y_pred_train, y_obs_train).add_suffix('_train')
                df_val = evaluation(y_pred_val, y_obs_val).add_suffix('_val')
                df_test = evaluation(y_pred_test, y_obs_test).add_suffix('_test')
    
                df_all_metrics = pd.concat([df_train, df_val, df_test], axis=1)
                df_all_metrics.index.name = length
    
                # Append df_all_metrics to the list
                df_all_metrics_list.append(df_all_metrics)
    
                # Calculate mean for all output step
                #mean_metrics = df_val.mean()
                mean_metrics = pd.concat([df_train.mean(), df_val.mean(), df_test.mean()])
                mean_metrics_row = pd.DataFrame(mean_metrics).T
                mean_metrics_row['MSE_val(loss)'] = mse
                mean_metrics_row['input_len'] = length
                #mean_metrics_row = mean_metrics_row[['input_len','MSE_val(loss)', 'MAE_val', 'RMSE_val', 'MAPE (%)_val']]
                mean_metrics_row = mean_metrics_row[['input_len', 'MSE_val(loss)', 'MAE_train', 'RMSE_train', 'MAPE (%)_train',
                                                 'MAE_val', 'RMSE_val', 'MAPE (%)_val',
                                                 'MAE_test', 'RMSE_test', 'MAPE (%)_test']]
    
                # Append to the list
                mean_metrics_rows.append(mean_metrics_row)
        else:
            logging.info(f"Run {run} has mse=NaN, skipping.")
            
    # Check if there are successful runs before proceeding
    if successful_runs:       
        # Concatenate all df_all_metrics into a single DataFrame with a new level for runs
        concatenated_all_metrics = pd.concat(df_all_metrics_list, keys=successful_runs, names=['Run', 'Time Step'])

        # Calculate the mean across runs for each metric and time step
        # This will group by 'Time Step' and calculate the mean of each metric across all runs
        aggregated_all_metrics_mean = concatenated_all_metrics.groupby('Time Step').mean()
        aggregated_all_metrics_mean.index.name = length

        logging.info("\n--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---")
        # Convert DataFrame to string
        aggregated_all_metrics_mean_str = aggregated_all_metrics_mean.to_string(index=False)    
        # Log the DataFrame
        logging.info("\n" + aggregated_all_metrics_mean_str)
        display(aggregated_all_metrics_mean)
    else:
        logging.info(f"No successful runs for input length {length}.")

    # Check if mean metrics were calculated
    if mean_metrics_rows:        
        # After all runs, create a DataFrame of mean metrics
        mean_metrics_df = pd.concat(mean_metrics_rows, ignore_index=True)
        # Calculate the mean of each metric across the 10 runs
        final_mean_metrics = mean_metrics_df.mean()

        # Create a DataFrame for the final mean metrics
        final_mean_metrics_df = pd.DataFrame(final_mean_metrics).T

        mean_metrics_list.append(final_mean_metrics_df)
    else:
        logging.info(f"No mean metrics calculated for input length {length}.")

mean_metrics_df = pd.concat(mean_metrics_list).reset_index(drop=True)
logging.info("\n--- Final Mean Metrics Across 10 Runs ---")
# Convert DataFrame to string
mean_metrics_df_str = mean_metrics_df.to_string(index=False)    
# Log the DataFrame
logging.info("\n" + mean_metrics_df_str)
mean_metrics_df

2025-01-13 03:13:04,999 - 336
2025-01-13 03:13:05,002 - 
--- Run 1 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 24: early stopping
Restoring model weights from the end of the best epoch: 4.


2025-01-13 03:17:00,921 - Validation loss: 0.02516
2025-01-13 03:17:00,924 - Run 1 has mse=0.025159712880849838, skipping.
2025-01-13 03:17:00,925 - 
--- Run 2 ---


Epoch 47: early stopping
Restoring model weights from the end of the best epoch: 27.


2025-01-13 03:24:43,796 - Validation loss: 0.00592
2025-01-13 03:24:54,346 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 27: early stopping
Restoring model weights from the end of the best epoch: 7.


2025-01-13 03:29:27,957 - Validation loss: 0.00692
2025-01-13 03:29:38,309 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 20: early stopping
Restoring model weights from the end of the best epoch: 1.


2025-01-13 03:32:59,647 - Validation loss: nan
2025-01-13 03:32:59,649 - Run 4 has mse=NaN, skipping.
2025-01-13 03:32:59,650 - 
--- Run 5 ---


Epoch 62: early stopping
Restoring model weights from the end of the best epoch: 42.


2025-01-13 03:43:08,017 - Validation loss: 0.00595
2025-01-13 03:43:19,156 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 48: early stopping
Restoring model weights from the end of the best epoch: 28.


2025-01-13 03:51:20,117 - Validation loss: 0.00606
2025-01-13 03:51:31,160 - 
--- Run 7 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 57: early stopping
Restoring model weights from the end of the best epoch: 37.


2025-01-13 04:01:16,919 - Validation loss: 0.00597
2025-01-13 04:01:28,173 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 76: early stopping
Restoring model weights from the end of the best epoch: 56.


2025-01-13 04:14:08,167 - Validation loss: 0.00570
2025-01-13 04:14:19,388 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 59: early stopping
Restoring model weights from the end of the best epoch: 39.


2025-01-13 04:24:20,845 - Validation loss: 0.00588
2025-01-13 04:24:31,735 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 43: early stopping
Restoring model weights from the end of the best epoch: 23.


2025-01-13 04:31:47,109 - Validation loss: 0.02186
2025-01-13 04:31:47,111 - Run 10 has mse=0.02186303399503231, skipping.
2025-01-13 04:31:47,117 - 
--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---
2025-01-13 04:31:47,122 - 
 MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
 23.414601   32.056683    1.101442e+11 21.955408 33.939292     42.513167 25.170847  36.277035      32.652209
 24.625599   34.198602    1.457155e+11 23.465655 37.767377     40.869351 27.769847  41.112069      34.148795
 25.430239   35.535308    1.716921e+11 24.243129 39.679370     42.261954 29.199529  43.690869      36.273759
 26.085867   36.548539    2.065791e+11 24.929341 41.015453     42.571067 30.414164  45.776318      38.001377
 26.803073   37.401393    2.346961e+11 25.638992 42.102602     46.178403 31.468251  47.434822      41.157452
 27.585411   38.158024    2.520001e+11 26.415916 42.841416     53.844137 32.017235  48.554717      44.8

MAE_train  RMSE_train  MAPE (%)_train    MAE_val   RMSE_val  \
336                                                                
1    23.414601   32.056683    1.101442e+11  21.955408  33.939292   
2    24.625599   34.198602    1.457155e+11  23.465655  37.767377   
3    25.430239   35.535308    1.716921e+11  24.243129  39.679370   
4    26.085867   36.548539    2.065791e+11  24.929341  41.015453   
5    26.803073   37.401393    2.346961e+11  25.638992  42.102602   
6    27.585411   38.158024    2.520001e+11  26.415916  42.841416   

     MAPE (%)_val   MAE_test  RMSE_test  MAPE (%)_test  
336                                                     
1       42.513167  25.170847  36.277035      32.652209  
2       40.869351  27.769847  41.112069      34.148795  
3       42.261954  29.199529  43.690869      36.273759  
4       42.571067  30.414164  45.776318      38.001377  
5       46.178403  31.468251  47.434822      41.157452  
6       53.844137  32.017235  48.554717      44.857846

2025-01-13 04:31:47,135 - 48
2025-01-13 04:31:47,136 - 
--- Run 1 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 37: early stopping
Restoring model weights from the end of the best epoch: 17.


2025-01-13 04:38:26,758 - Validation loss: 0.00580
2025-01-13 04:38:32,531 - 
--- Run 2 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 40: early stopping
Restoring model weights from the end of the best epoch: 20.


2025-01-13 04:45:55,456 - Validation loss: 0.00587
2025-01-13 04:46:01,470 - 
--- Run 3 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 34: early stopping
Restoring model weights from the end of the best epoch: 14.


2025-01-13 04:51:48,721 - Validation loss: 0.00564
2025-01-13 04:51:53,699 - 
--- Run 4 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 61: early stopping
Restoring model weights from the end of the best epoch: 41.


2025-01-13 05:03:18,061 - Validation loss: 0.01326
2025-01-13 05:03:18,065 - Run 4 has mse=0.013257408514618874, skipping.
2025-01-13 05:03:18,066 - 
--- Run 5 ---


Epoch 48: early stopping
Restoring model weights from the end of the best epoch: 28.


2025-01-13 05:12:01,916 - Validation loss: 0.00559
2025-01-13 05:12:07,962 - 
--- Run 6 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 23: early stopping
Restoring model weights from the end of the best epoch: 3.


2025-01-13 05:16:19,827 - Validation loss: 0.01967
2025-01-13 05:16:19,830 - Run 6 has mse=0.019665325060486794, skipping.
2025-01-13 05:16:19,832 - 
--- Run 7 ---


Epoch 53: early stopping
Restoring model weights from the end of the best epoch: 33.


2025-01-13 05:25:36,744 - Validation loss: 0.00552
2025-01-13 05:25:42,494 - 
--- Run 8 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 44: early stopping
Restoring model weights from the end of the best epoch: 24.


2025-01-13 05:33:29,040 - Validation loss: 0.00570
2025-01-13 05:33:34,298 - 
--- Run 9 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 40: early stopping
Restoring model weights from the end of the best epoch: 20.


2025-01-13 05:40:22,334 - Validation loss: 0.00567
2025-01-13 05:40:27,751 - 
--- Run 10 ---
/home/yl2672496l/Yue/code/jupyter_env/lib/python3.12/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 41: early stopping
Restoring model weights from the end of the best epoch: 21.


2025-01-13 05:47:27,315 - Validation loss: 0.00574
2025-01-13 05:47:33,103 - 
--- Aggregated Mean of All Metrics Across 10(Successful) Runs ---
2025-01-13 05:47:33,106 - 
 MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
 22.803459   32.110652    7.577846e+10 20.646787 32.056396     37.303585 23.005045  33.472334      26.451181
 25.319244   35.661641    1.134349e+11 23.339565 36.761639     42.026295 26.884089  39.335143      30.953135
 26.278163   37.153974    1.450604e+11 24.232111 38.727691     43.544529 28.543338  41.877545      33.327479
 27.073762   37.887404    1.751645e+11 25.229256 39.809423     49.778754 29.707718  43.337379      37.810941
 28.000700   39.036776    1.986458e+11 26.095277 40.854141     55.257476 30.785630  44.917477      41.768766
 28.939585   40.100789    2.186869e+11 26.985615 41.805527     61.243478 31.592956  46.126067      44.771984


MAE_train  RMSE_train  MAPE (%)_train    MAE_val   RMSE_val  MAPE (%)_val  \
48                                                                              
1   22.803459   32.110652    7.577846e+10  20.646787  32.056396     37.303585   
2   25.319244   35.661641    1.134349e+11  23.339565  36.761639     42.026295   
3   26.278163   37.153974    1.450604e+11  24.232111  38.727691     43.544529   
4   27.073762   37.887404    1.751645e+11  25.229256  39.809423     49.778754   
5   28.000700   39.036776    1.986458e+11  26.095277  40.854141     55.257476   
6   28.939585   40.100789    2.186869e+11  26.985615  41.805527     61.243478   

     MAE_test  RMSE_test  MAPE (%)_test  
48                                       
1   23.005045  33.472334      26.451181  
2   26.884089  39.335143      30.953135  
3   28.543338  41.877545      33.327479  
4   29.707718  43.337379      37.810941  
5   30.785630  44.917477      41.768766  
6   31.592956  46.126067      44.771984

2025-01-13 05:47:33,116 - 
--- Final Mean Metrics Across 10 Runs ---
2025-01-13 05:47:33,120 - 
 input_len  MSE_val(loss)  MAE_train  RMSE_train  MAPE (%)_train   MAE_val  RMSE_val  MAPE (%)_val  MAE_test  RMSE_test  MAPE (%)_test
     336.0       0.006057  25.657465   35.649758    1.868045e+11 24.441407 39.557585     44.706346 29.339979  43.807638      37.848573
      48.0       0.005691  26.402485   36.991873    1.544618e+11 24.421435 38.335803     48.192353 28.419796  41.510991      35.847248


input_len  MSE_val(loss)  MAE_train  RMSE_train  MAPE (%)_train    MAE_val  \
0      336.0       0.006057  25.657465   35.649758    1.868045e+11  24.441407   
1       48.0       0.005691  26.402485   36.991873    1.544618e+11  24.421435   

    RMSE_val  MAPE (%)_val   MAE_test  RMSE_test  MAPE (%)_test  
0  39.557585     44.706346  29.339979  43.807638      37.848573  
1  38.335803     48.192353  28.419796  41.510991      35.847248